src/main/scala/tech/cryptonomic/conseil/routes/DataHelpers.scala

```diff
@@ -20,11 +20,26 @@ trait DataHelpers extends QueryStringListsServer with Validation with akkahttp.s
   import io.circe.syntax._
 
   /** Function validating request for the query endpoint */
-  override def validated[A](response: A => Route, invalidDocs: Documentation): Either[List[QueryValidationError], A] => Route = {
+  override def validated(response: QueryResponseWithOutput => Route, invalidDocs: Documentation): Either[List[QueryValidationError], QueryResponseWithOutput] => Route = {
     case Left(errors) =>
       complete(StatusCodes.BadRequest -> s"Errors: \n${errors.mkString("\n")}")
-    case Right(success) =>
+    case Right(success@QueryResponseWithOutput(_, OutputType.json)) =>
       response(success)
+    case Right(QueryResponseWithOutput(queryResponse, OutputType.csv)) =>
+      complete(StatusCodes.OK -> makeCsv(queryResponse))
```

Comment:
Is there a way to make this choice implicit in the way the response is marshsalled?

To avoid explicitly duplicate the response with different calls (response(success) vs. complete(...))?

Also, what about the response mime-type? Shouldn't it appear somewhere here?

Reply:
Yes it was wrong, I think I needed to take a step back during the weekend and think about it once again. Now it correctly returns `text/csv`!

Reply:
And regarding explicitly choosing the way the responses are marshalled -> I could't find a reasonable way to make this choice implicit.